src/librbd/Utils.cc

```diff
@@ -147,9 +163,10 @@ int create_ioctx(librados::IoCtx& src_io_ctx, const std::string& pool_desc,
     return r;
   }
 
-  dst_io_ctx->set_namespace(
-    pool_namespace ? *pool_namespace : src_io_ctx.get_namespace());
-  if (src_io_ctx.get_pool_full_try()) {
+  if (pool_namespace) {
+    dst_io_ctx->set_namespace(*pool_namespace);
+  }
```

Comment:
This does not look correct. Previously not specified namespace meant that we want the dst namespace the same as the src namespace (and I recall we had code that relied on that behavior). After your change, the not specified namespace means "use the currently set dst namespace".

Reply:
`create_ioctx` logic did not change.
For the new `create_ioctx2`, used only by `NativeFormat.cc` - you're right. I will make a fix to `NativeFormat.cc` to set `pool_namespace` to the src namespace if it's empty.

Reply:
Actually, I take that back. I think that `NativeFormat.cc` is already fine as it is.
Notice that if `pool_namespace_val.type() == json_spirit::null_type`, then `m_pool_namespace = m_image_ctx->md_ctx.get_namespace();`

Reply:
I don't like that `create_ioctx` and `create_ioctx2` have different semanic. 

I.e. in `create_ioctx`, if `pool_namespace` is not specified it means "to set the destination namespace to the same value as the source namespace". While in `create_ioctx2` if `pool_namespace` is not specified it means "not to set the destination namespace at all (use the default one)". And this is confusing. Even if the current code works correctly it may be misused in the future.

Also, why don't use the same `create_ioctx` name for the new function, just a different signature? Why introducing `create_ioctx2` name?

Reply:
But `create_ioctx2` does not have any source ioctx to copy the namespace from.
Anyhow, do you have a concrete suggestion to better this?

Reply:
Make `pool_namespace` non-optional in your `create_ioctx2`?